feat: harness collects profiler metrics [DET-5061]

armandmcqueen opened the pull request:
## Description

Moving from PR #2073 so that branch isn't in personal repo and can be used as a feature branch

## Test Plan

<!---
Describe the situations in which you've tested your change, and/or a screenshot as appropriate.
Reviewers may ask questions of this test plan to ensure adequate manual coverage of changes.
-->


## Commentary (optional)

<!---
Use this section of your description to add context to the PR. Could be for particularly
tricky bits of code that could use extra scrutiny, historical context useful for reviewers, etc.
You may intentionally leave this section blank and remove the title.
--->


## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.


<!---
## Title

Example title: "docs: tweak recommended "pip install" usage".

Specifically, this title should contain a type and a description
of the change being made:

User-facing change types:
- docs: docs-only change
- feat: new user-facing feature
- fix: bug fix
- perf: performance improvement

Internal change types:
- build: build system change (anything in a `Makefile`, mostly)
- chore: any internal change not covered by another type
- ci: anything that touches `.circleci`
- refactor: internal refactor
- style: style change
- test: new tests

See https://www.conventionalcommits.org/en/v1.0.0/ for background.

The first line should also:
- be at most 89 characters long
- contain a description that is at most 72 characters long
- not end with sentence-ending punctuation
- start (after the type) with a lowercase imperative ("add", "fix")
-->

armandmcqueen commented:
Note from old PR on how to disable GPU metrics if pynvml isn't available or there are no GPUs:
```
try:
    import pynvml
    pynvml.nvmlInit()
except ModuleNotFoundError:
    pass
except pynvml.NVMLError_LibraryNotFound:
    pass
```

rb-determined-ai commented on harness/determined/profiler.py:
```diff
@@ -0,0 +1,517 @@
+        # Set up the thread responsible for making API calls
+        self.send_queue = queue.Queue()
+        self.sender_thread = ProfilerSenderThread(self.send_queue, self.master_url)
+        self.sender_thread.start()
```
Normally, python threds have a distinct "start/stop" lifetime which is a subset of their "init/gc" lifetime.  Or in other words, normally threads aren't started in `__init__()`, only created there.

Examples:
```python
# threading library
t = threading.Thread(target=my_func)
t.start()
t.join()
# gc happens later
```

Can we have a `ProfilingAgent.start()` method that starts both threads?  And a matching `ProfilingAgent.close()` method that (idempotently) stops/joins both threads?

Then the step after that is to write an `__enter__()` method that does `self.start(): return self` and an `__exit__()` that does `self.close` and you'll have a real nice API that will always shut down the threads it started, even in the case of uncaught exceptions.

armandmcqueen replied:
So if you don't mind, let's defer this problem. There is a PR that follows up on this one that hooks it into the TrialControllers. The current proposal is to use a context manager so we can move the start into that. If we decide not to use a context manager, then we can add a start() function

rb-determined-ai replied:
Sure, I'm ok deferring this.  Although I think we should use a context manager and the context manager should call start() in it's `__enter__()` function.  That way the primary usage pattern is super easy, but you still give the flexibility to call it in whatever strange calling pattern you want (such as we needed to integrate it into a Keras Callback).

armandmcqueen replied:
Fixed

rb-determined-ai commented on harness/determined/profiler.py:
```diff
@@ -0,0 +1,517 @@
+        # Track duration to stop collecting after 5 minutes. start_time also serves as
+        # the indicator that collection has begun.
+        self.start_time = None
+        # TODO: Currently we only check for shutdown in update_batch_idx(). This is a problem when there
+        #       is a long period of time between batch_idx being updated (such as during validation)
+        #       because it won't shut down until validation completes. We probably need to create a
+        #       new thread to enforce the timeout correctly.
```
Why not just explicity shut down the collection when we transition from training -> validation?  Or hardcode the timeout into the collection thread rather than ProfilerAgent?

rb-determined-ai commented on harness/determined/profiler.py:
```diff
@@ -0,0 +1,517 @@
+        self.has_started = False
+        self.has_finished = False
+        self.max_collection_seconds = 300
+        self.shutdown_timer = threading.Timer(self.max_collection_seconds, self._end_collection)
```
Kicking off a thread just for this seems like overkill when you could just start the SysMetricCollectorThread with a parameter of 300, so that it just dies if it collects that far.

That would have the extra benefit that you wouldn't the shutdown_lock since `_end_collection` would only get called from the main thread.

armandmcqueen replied:
Collecting Timings and batching them will probably require a second thread to run in parallel to the SysMetricsCollector thread that should also shut down after 5 minutes. In that case it is cleaner to have a single shutdown timer. If not, the cost of having the shutdown timer really isn't that high.

rb-determined-ai replied:
Sure, but why would you not just have the collector thread always send a Shutdown message to the sender thread?  That way you really do only have one timer in the system (I'm counting the collector thread as a form of a timer during its normal operation).  You also would get that totally for free.

I think what I am more concerned about here is that there's no way to close down this Timer thread.  It's not that it's *that* expensive (though it is creating an entire thread where a simple `if` statement would suffice); it's not that it's *that* risky to let it just sit around for 5 minutes in the background.  But to me it's definitely an "ease of reasoning about" issue.

If there is one pattern in concurrency that makes concurrency easy to reason about, it's [structured concurrency](http://libdill.org/structured-concurrency.html), which is basically the rule that anything that spins off subroutines will always clean up the subroutines before exiting.  It's easy to reason about because when the parent object is closed, you know that all child objects (threads, subprocesses, whatever) are also closed.  Also, it is easy to test because if you ever run two similar tests in a row you don't have leftovers from one test stepping on the next one.

We didn't used to practice structured concurrency in the harness, but we would run into weird bugs where e.g. unit tests would OOM a GPU if you tried to run the whole unit test suite all at once (https://determinedai.atlassian.net/browse/DET-2026) or there was indeterministic shutdown order between when ZMQ sockets got closed vs when the interpreter shut down, which left cryptic error messages everywhere.

armandmcqueen replied:
You have the Profiler wrapper, then you have two child threads the SysMetricCollector and the TimingsBatcher, both of which depend on a third child thread, SenderThread. Putting the SysMetricCollector in charge of controlling timeouts and the SenderThread doesn't make any sense to me - then you need to synchronize across two threads to shut down correctly when the SysMetricCollector and TimingsBatcher threads otherwise wouldn't need to communicate at all. 

If you insist, I will implement a custom Timer class that can be manually cleaned up, but I feel like that is spending time solving a problem that doesn't really occur. Is the specific case you are concerned about is when you write a test that run the harnesses multiple times inside the same process so the timer persists across harness executions?

armandmcqueen replied:
I'll write a custom Timer class. That seems like the solution that will make us both happy

armandmcqueen replied:
We talked in person. Sticking with the Timer thread for now (custom so it can follows structured concurrency)  so we don't need to introduce communication between the SysMetricCollector and TimingsBatcher threads.

rb-determined-ai commented on harness/determined/profiler.py:
```diff
@@ -0,0 +1,517 @@
+        max batch idx being exceeded or due to timeout, so the function needs to be threadsafe.
+        """
+        # Make _end_collection idempotent
+        with self.shutdown_lock:
+            if self.has_finished:
+                return
+
+            # Shut down in reverse creation order
+            self.sys_metric_collector_thread.kill()
+            self.sys_metric_collector_thread.join()
+
+            self.sender_thread.kill()
+            self.sender_thread.join()
+
+            self.has_finished = True
```
Do you actually need the shutdown lock and the has_finished check?  I think the kill() calls and the join() are basically already idempotent (other than extra dummy messages getting put into queues... pretty harmless)

armandmcqueen replied:
I think using a single lock is cleaner, easier to reason about and it comes at no meaningful cost. `kill` doesn't need to be idempotent, that's just a benefit of how it's currently implemented, which could easily change in the future. I'd rather write code that handles the concurrency explicitly in the function that needs to be threadsafe rather than relying on implementation details of a function that has no reason to remain idempotent.

rb-determined-ai replied:
Honestly I think any differences between cleanliness and ease of reasoning between either the lock solution or the queue solution are pretty tiny.

I guess that's not really what I'm concerned about.  I'll respond in more detail on the other comment.

shiyuann commented on harness/determined/profiler.py:
```diff
@@ -0,0 +1,559 @@
+    def is_enabled(self) -> bool:
+        return self.has_started and not self.has_finished
+
+    def update_batch_idx(self, new_batch_idx: int) -> None:
```
q: should this be called for each batch? If so, probably add a check here.

armandmcqueen replied:
That's an interesting point. It should be called for every batch, but I'm not certain if it's incorrect if it isn't, particularly when we think about it being used outside of our harness/with the generic APIs. We'll talk about this at standup.

shiyuann replied:
It looks like `new_batch_idx` should be at least larger than or equal to `current_batch_idx`.

armandmcqueen replied:
Fixed

shiyuann commented on harness/determined/profiler.py:
```diff
@@ -0,0 +1,559 @@
+                batch_start_time = time.time()
+
+
+# This is a thread that exists solely so that we can make API calls without blocking
```
nit: can you move this to the docstring of this class.

armandmcqueen replied:
Fixed

shiyuann commented on harness/determined/profiler.py:
```diff
@@ -0,0 +1,559 @@
+                except queue.Empty:
+                    pass
+
+            next_collection += self.MEASUREMENT_INTERVAL
```
q: should this be
```suggestion
            next_collection = time.now() + self.MEASUREMENT_INTERVAL
```
in case `MEASUREMENT_INTERVAL` is very small?

 In practice, `next_collection - now` could be negative.

rb-determined-ai replied:
This was my idea actually.  This keeps the interval as close to regular as possible for normal intervals.  If the interval is too small, it exhibits the behavior of collecting as fast as technically possible, which seems like the right thing to do if the interval is too small.

shiyuann replied:
If sleep time can be negative, it seems fine to me.

armandmcqueen replied:
If the timeout is negative, the code will raise an exception. Good catch. I changed it so that the sleep time is never less than 0.

armandmcqueen replied:
Fixed

shiyuann commented on harness/determined/profiler.py:
```diff
@@ -0,0 +1,559 @@
+    """
+
+    FLUSH_INTERVAL = 10  # How often to make API calls
+    MEASUREMENT_INTERVAL = 0.1
```
q: should this be configurable?

armandmcqueen replied:
It should be configurable by us in the future, but it doesn't not need to be configurable by users

shiyuann commented on harness/determined/profiler.py:
```diff
@@ -0,0 +1,559 @@
+
+        trial_profiler_metrics_batches = []
+        for metric_name in self.batch.keys():
+            # TODO: Don't forget to include GPU Memory
```
style: generally, we need to include a ticket number or this would be resolved in the same PR otherwise it wouldn't be tracked.

armandmcqueen replied:
I just missed these TODOs. They need to be fixed before this PR is merged and I'll fix that now. Nice catch!

shiyuann commented on harness/determined/profiler.py:
```diff
@@ -0,0 +1,559 @@
+                )
+
+            # GPU Metrics need to be grouped by GPU UUID
+            # TODO: Don't forget to include GPU Memory
```
same

shiyuann commented on harness/determined/profiler.py:
```diff
@@ -0,0 +1,559 @@
+    def activate(self) -> None:
+        self.control_queue.put(StartMessage())
+
+    def kill(self) -> None:
```
This is actually "stop" rather than "kill".

```suggestion
    def stop(self) -> None:
```

armandmcqueen replied:
Why is this "stop" rather than "kill"? It doesn't complete active work before shutting down, it just drops any in-progress batches, which seems more like "kill" to me.

shiyuann reviewed (commented):
looks good overall. So I left some minor comments.

armandmcqueen commented on harness/determined/exec/worker_process.py:
```diff
@@ -27,6 +28,9 @@ def main() -> None:
 
     config_logging(worker_process_env)
 
+    # API code expects credential to be available as an environment variable
```
This is just how authentication works in our `common/api` code AFAIU. While it's not ideal, I think that refactoring the python API code to not rely on environment variables is out of scope of this PR. Non-DTrain already relies on this envvar-based pattern IIUC, this code is just fixing the bug where the credentials aren't passed around correctly in DTrain.

shiyuann replied:
@rb-determined-ai did quite a lot of work getting rid of using environment variables to pass around information. If he is fine with this, let's create a ticket to refactor this later.

rb-determined-ai replied:
I'm fine with this for now.  But yeah, let's make a ticket.

rb-determined-ai replied:
@armandmcqueen can you make the ticket?

armandmcqueen commented on harness/determined/profiler.py:
```diff
@@ -0,0 +1,559 @@
+    """
+    Agent that collects metrics and sends them to the master. It has:
+    - a thread to send data to the Master API (sender_thread)
+    - a thread to collect System Metrics and periodically flush to
```
It's using capitalization to emphasize that these are concepts in Determined and not just descriptions.

shiyuann replied:
I see. I couldn't find them in the docs or the codebase. Can you define them in this file or somewhere else?

armandmcqueen replied:
We'll add them in the docs. I updated the docs ticket to make sure we do that.

armandmcqueen commented on harness/determined/_execution.py:
```diff
@@ -126,7 +126,9 @@ def _make_local_execution_env(
         det_trial_unique_port_offset=0,
         det_trial_runner_network_interface=constants.AUTO_DETECT_TRIAL_RUNNER_NETWORK_INTERFACE,
         det_trial_id="",
+        det_agent_id="local",
```
I'll make it empty. Out of curiosity, why does it matter?

armandmcqueen replied:
Fixed

armandmcqueen commented:
Updated the PR a good bit since the last reviews:

- Addressed most comments
- We now check if we pynvml is functioning so that we don't get failures if pynmvl isn't available/there aren't GPUs. I don't love the implementation, but a full-blown wrapper around pynvml like PolyHorovod seemed like overkill
- Added more inputs to the ProfilingAgent such as global rank and local rank. 
- Separated the concept of 'enabled' from 'active'. There is more detail on this in the design doc
- Added `start` and `end` functions as well as a context manager for the ProfilingAgent
- Added a custom timer that can be cleaned up if we need to exit before it completes

I also added a design doc to the codebase, which I feel like people might object to. My thinking is that in 6 months or a year when I don't remember the details of this code or when someone else needs to work on this code, it's really useful to have some docs to remember how it works and why it was written this way.  When I inline substantial documentation, the problem is that the implementation and the documentation drift apart over time and the documentation becomes misleading. One pattern I've seen is to include immutable design docs directly in the codebase with a date attached. This way it doesn't need to be updated, but it is still useful as a reference. I'm curious what people think about the idea

rb-determined-ai commented on harness/determined/profiler.py:
```diff
@@ -0,0 +1,666 @@
+        # TODO: Add new timing to TimingBatcher
+
+
+class CustomTimer(threading.Thread):
```
PreemptibleTimer?

armandmcqueen replied:
Fixed

rb-determined-ai commented on harness/determined/profiler.py:
```diff
@@ -0,0 +1,666 @@
+    ):
+        self.duration = duration
+        self.callback = callback
+        self.callback_args = callback_args if callback_args is not None else []
```
callback_args seems unnecessary; I think args should be bound into a closure.  `labmda: my_fn(my_args)` is pretty common in python.

armandmcqueen replied:
Fixed - I just removed the concept of args altogether since it's just dead code now

rb-determined-ai commented on harness/determined/profiler_design_april_2021.md:
```diff
@@ -0,0 +1,114 @@
+# Profiler Design Notes (April 2021)
```
Is the plan to commit this into the codebase?

rb-determined-ai replied:
I see your comment now.  I'm not opposed to inline developer docs on principle (I've written [some](https://github.com/determined-ai/determined/blob/master/harness/determined/pytorch/README.md) myself).

But in this case I don't think the profiling system here is hard to understand, or convoluted, or presents a complex API to justify committing the docs to the repository.  Honestly I think the docstring that is already in ProfilingAgent's sums the system up plenty well.

armandmcqueen replied:
Removed it.

rb-determined-ai commented on harness/setup.py:
```diff
@@ -55,6 +55,7 @@
         "docker-compose>=1.13.0",
         "tqdm",
         "appdirs",
+        "backoff",
```
I'm flagging this as requiring changes to task environments so we are not surprised on release-party day.

armandmcqueen replied:
I think this means that after this PR is merged and prior to the next release, I need to make sure that the task envs is rebuilt by adding any commit. Is that correct?

rb-determined-ai replied:
Yes.

shiyuann reviewed (commented):
The current state looks good. The only things are the TODOs are not attached with ticket numbers so I'll review them after you fill them up.

armandmcqueen commented on harness/determined/profiler.py:
```diff
@@ -0,0 +1,666 @@
+
+        # Check if we should stop collecting metrics due to batch idx being exceeded
+        if (
+            self.is_active
+            and self.end_after_batch is not None
```
I think we have to keep the second line so I'm going to leave this as is

armandmcqueen commented on harness/determined/profiler.py:
```diff
@@ -0,0 +1,666 @@
+
+LOG_NAMESPACE = "determined-profiler"
+
+SHOULD_PROFILE_GPUS = None
```
Fixed

armandmcqueen commented on harness/determined/profiler.py:
```diff
@@ -0,0 +1,666 @@
+        self.callback = callback
+        self.callback_args = callback_args if callback_args is not None else []
+        self._timer_has_begun = False
+        self.control_queue: "queue.Queue[Union['StartMessage', 'ShutdownMessage']]" = queue.Queue()
```
Fixed

armandmcqueen commented on harness/determined/profiler.py:
```diff
@@ -0,0 +1,666 @@
+        # Check if we should start collecting metrics
+        if (
+            not self.is_active
+            and not self.has_finished
```
Fixed

armandmcqueen commented:
All comments have been addressed except for the ones that I left unresolved.

rb-determined-ai reviewed (approved):
nice work!
